marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs

```diff
@@ -0,0 +1,734 @@
+  , "d3e96ebec374f10c072b403e6f07ffe533f0bbd13d9dbd01f24b9ac78ec2279f"
+  , "a66e0ca2842dc8cd873af85e68bdfcf452b674fd7b5f603782c7d4864b72ec4f"
+  , "0b02d9e23e96e2f8f5947e6ac226add43359c34afc86fe2970482d0b60b7eaf8"
+  , "22e8cdd799d7c7de2475bc076517fb03855f229c146d6c477d322490905b80b6"
+  , "2a2bd25ce895f9235f8421cc500d342af5b07315a76b0bb3ed9bfab0c15b6e79"
+  , "cfbd260b5bacf6bf063c74ec3eae9438b386a58f24ad72b4603fab2e22d7219c"
+  , "e1b8c48af9f4dfabf92b8500978e9e3af06c6da9139ee489973ceb42ec875e0b"
+  , "96dc5b97f8a7de530533d6b812057a5b878298ff14425abefa0da43ca037acc7"
+  , "224da92658a1a35821f9f21324ed7b9aeb09ca023660ee1ae168ac1c15bea458"
+  , "9e39fff27a033125c176a3bca10e7a5bab471fc883fa839e5b6569cbc1beb1a6"
+  , "4ef07b4a7f5261f8f7facc70834789f37da50f7dcb1b5197d6490c0d114b91a0"
+  , "f591e43295c0a15883cda658daeb8815ae9f667f150cd98704a0eac2ab1a4bde"
+  , "79898783ed4287d72324dd979d275b807e589e5a40dd72b3fe03e295bf137020"
+  , "bf3dd6f71e21a6e65c6b0ab444b45dcc854b10d5f8aa431cc29a63349201cb48"
+  , "11837b6d83224a3b51b09f40fd341cf13a1149a29f2525e9c29298a585c8f265"
+  , "76acb519350d25a781b0345171e3da6840d65fe28141bece75c811eb1e044394"
+  , "a72d43b5a10e3c7b750c97283bae5fb0aa36db7b44b0604e815550ff7a034f4e"
+  , "0046519ac36c44ef2739a5b17b6d49475828f11a02037d3dbbbb1d663409a82e"
+  , "5e631d10f0019395570d09808857b2c9fa9b3bbdeef7081e6d0a62786b77215e"
+  , "7c05d520230661de5bed21dabcbbc211e49d8dba96b604360e600480ae1d9d22"
+  , "2b16eb1a58206124ebff59f92474fe7853e5fe2af813717e2c049239ef994f29"
+  , "f0504414bf68edfa27adb71fadae6950fb02292018909cb6c700ea55fa1eb5f7"
+  , "a146de3c069760deeacb8132b979b914ee10e9c2282b245db1c7f41d1e383d04"
+  , "4fa9106c10efbc65a9c8b6b92527c6222c2f970b9258096e8f4ce6d6478bec92"
+  , "4e65f9fff17d94c5415d0086d9ec573f2d3b66c42375f978927395658e3e641a"
+  , "577d3f0111f5bb67b1a0794bdfb596de9998ce249fde1fc5ba3ebd6d820d237e"
+  , "1d6ca23dd7348e2c631200c900ad7f0d49d63cdf9d3291f1791f03c3c699b252"
+  , "339554c1f558b1b98b10524d4104453eb54fae22b7dfb118caa850efe069f4f3"
+  , "2ecebb248671fbe25a8bfc757da41e720fe1b4425eaa82f97abff6a6d057e706"
+  , "5853f8a0c295007e1e89287f578158c0db8f57e5a256b15d56331620789a5ff5"
+  , "1cad64602aef420c7386ac16ada8260ce609f3dceacd498de8ae4a186f83f958"
+  ]
+
+genMintScriptWitness :: Gen (ScriptWitness WitCtxMint BabbageEra)
+genMintScriptWitness = oneof
+  [ PlutusScriptWitness
+      PlutusScriptV1InBabbage
+      PlutusScriptV1
+      <$> (hedgehog $ PScript <$> genPlutusScript PlutusScriptV1)
+      <*> pure NoScriptDatumForMint
+      <*> hedgehog genScriptData
+      <*> (ExecutionUnits <$> (fromIntegral @Word32 <$> arbitrary) <*> (fromIntegral @Word32 <$> arbitrary))
+  ]
+
+genRoleToken :: Gen Chain.AssetId
+genRoleToken = Chain.AssetId
+  <$> (hedgehog $ fromCardanoPolicyId . PolicyId <$> genScriptHash)
+  <*> genRole
+
+-- NOTE just a random list of names that won't conflict with anything generated
+-- by Gen.Cardano.Api.Typed
+genRole :: Gen Chain.TokenName
+genRole = elements
```

Comment:
Someday we should unify the `Gen` instances here (and in other packages) with those in `marlowe-test`, so we have less code to maintain.

Reply:
Somehow, yes - although it's a bit tricky because context matters (in the case of Chain.TokenName, it could be a random token or a role token)